### ISSUE TITLE
Attention maps 可视化代码

### ISSUE THREAD

**Issue by mickiefan** [Mr-Bigworth/MMCA#8]
作者您好，请问可以给一下Attention maps可视化的代码吗？

**Comment by Mr-Bigworth** [comment]
可以的，是指基于MMCA里的attention的可视化过程吗，你的attention是怎样实现的哇。我论文里可视化实际上是沿hw*hw的attention map（softmax之前）的第一个维度求和并沿head求平均，本质上是衡量输入的token对于下一层的贡献度。

**Comment by mickiefan** [comment]
> 可以的，是指基于MMCA里的attention的可视化过程吗，你的attention是怎样实现的哇。我论文里可视化实际上是沿hw*hw的attention map（softmax之前）的第一个维度求和并沿head求平均，本质上是衡量输入的token对于下一层的贡献度。

对的，是指基于MMCA里的attention的可视化过程。我目前还没有做太多的改动，想先画出可视化的图看下效果，看下这个模型在我的任务里效果怎么样，所以想问下可视化的代码。

**Comment by Mr-Bigworth** [comment]
> > 可以的，是指基于MMCA里的attention的可视化过程吗，你的attention是怎样实现的哇。我论文里可视化实际上是沿hw*hw的attention map（softmax之前）的第一个维度求和并沿head求平均，本质上是衡量输入的token对于下一层的贡献度。
> 
> 对的，是指基于MMCA里的attention的可视化过程。我目前还没有做太多的改动，想先画出可视化的图看下效果，看下这个模型在我的任务里效果怎么样，所以想问下可视化的代码。

我找找，稍等

**Comment by Mr-Bigworth** [comment]
> > 可以的，是指基于MMCA里的attention的可视化过程吗，你的attention是怎样实现的哇。我论文里可视化实际上是沿hw*hw的attention map（softmax之前）的第一个维度求和并沿head求平均，本质上是衡量输入的token对于下一层的贡献度。
> 
> 对的，是指基于MMCA里的attention的可视化过程。我目前还没有做太多的改动，想先画出可视化的图看下效果，看下这个模型在我的任务里效果怎么样，所以想问下可视化的代码。

时间有点久了，不过你应该可以用这个代码进行可视化，在MMCA里因为640 * 640的输入最后会在6层Transformer里是20 * 20个token，你在Transformer encoder的视觉自注意力里debug的时候用它可视化softmax前的attn，位置是[visual_model](https://github.com/Mr-Bigworth/MMCA/tree/main/models_mmca_vector_based/visual_model)
/transformer.py line 288，进到torch库里之后应该是_scaled_dot_product_attention这个函数里（因为在这里可以可视化softmax前的部分）
否则的话你可以通过修改
src2, attn = self.self_attn(q, k, value=src, attn_mask=src_mask,
                      key_padding_mask=src_key_padding_mask) 
来得到attn，但这样的attn是head做了平均并且经过softmax的，可以参考torch的文档。

以下是直接在torch源码里可视化
## 可视化视觉注意力
kk = attn.mean(dim=0).mean(dim=0) #使用mean
#kk = attn.sum(dim=0).sum(dim=0)  #使用sum
kk = kk.reshape(20,20)
import cv2
import matplotlib.pyplot as plt
import numpy as np
kk = kk.cpu().detach().numpy()
plt.imshow(kk, cmap=plt.cm.jet)
plt.colorbar()
plt.show()

**Comment by mickiefan** [comment]
好的好的，感谢🙏🙏
